solutions/USACO open16-splitting-the-field.mdx

```diff
@@ -92,262 +92,64 @@ From the analysis:
+		freopen((name+".out").c_str(), "w", stdout);
+	}
 }
 
-// OUTPUT
-tcT> void pr(T x) { cout << ts(x); }
-tcTUU> void pr(const T& t, const U&... u) { 
-	pr(t); pr(u...); }
-void ps() { pr("\n"); } // print w/ spaces
-tcTUU> void ps(const T& t, const U&... u) { 
-	pr(t); if (sizeof...(u)) pr(" "); ps(u...); }
-
-// DEBUG
-void DBG() { cerr << "]" << endl; }
-tcTUU> void DBG(const T& t, const U&... u) {
-	cerr << ts(t); if (sizeof...(u)) cerr << ", ";
-	DBG(u...); }
-#ifdef LOCAL // compile with -DLOCAL, chk -> fake assert
-	#define dbg(...) cerr << "Line(" << __LINE__ << ") -> [" << #__VA_ARGS__ << "]: [", DBG(__VA_ARGS__)
-	#define chk(...) if (!(__VA_ARGS__)) cerr << "Line(" << __LINE__ << ") -> function(" \
-		 << __FUNCTION__  << ") -> CHK FAILED: (" << #__VA_ARGS__ << ")" << "\n", exit(0);
-#else
-	#define dbg(...) 0
-	#define chk(...) 0
-#endif
-
-void setPrec() { cout << fixed << setprecision(15); }
-void unsyncIO() { cin.tie(0)->sync_with_stdio(0); }
-// FILE I/O
-void setIn(str s) { freopen(s.c_str(),"r",stdin); }
-void setOut(str s) { freopen(s.c_str(),"w",stdout); }
-void setIO(str s = "") {
-	unsyncIO(); setPrec();
-	// cin.exceptions(cin.failbit); 
-	// throws exception when do smth illegal
-	// ex. try to read letter into int
-	if (sz(s)) setIn(s+".in"), setOut(s+".out"); // for USACO
-}
+//constant initialization
+const ll INF = 1e18; //Not too close to LLONG_MAX
 
+//variables used for the current problem
 int N;
 ll ans = 0;
-vpi v;
+vector<pi> v;
 
 void tri() {
 	sort(all(v));
-	vpi lef(N), rig(N);
-	auto comb = [](pi a, int b) -> pi { return {min(a.f,b),max(a.s,b)}; };
+	vector<pi> lef(N), rig(N);
+	auto comb = [](pi a, int b) -> pi { return mp(min(a.f,b),max(a.s,b)); };
 	lef[0] = {v[0].s,v[0].s};
-	FOR(i,1,N) lef[i] = comb(lef[i-1],v[i].s);
-	rig[N-1] = {v[N-1].s,v[N-1].s};
-	R0F(i,N-1) rig[i] = comb(rig[i+1],v[i].s);
-	ll area = (ll)(v.bk.f-v.ft.f)*(lef.bk.s-lef.bk.f);
-	ll bes = INF;
-	F0R(i,N-1) if (v[i].f != v[i+1].f)
-		ckmin(bes,(ll)(v[i].f-v[0].f)*(lef[i].s-lef[i].f)
-			+(ll)(v.bk.f-v[i+1].f)*(rig[i+1].s-rig[i+1].f));
-	ckmax(ans,area-bes);
+	for (int i = 1; i < N; ++i) lef[i] = comb(lef[i-1],v[i].s);
+	rig[N-1] = mp(v[N-1].s,v[N-1].s);
+	for (int i = N-2; i >= 0; --i) rig[i] = comb(rig[i+1],v[i].s);
+	ll area = (ll)(v.back().f-v.front().f)*(lef.back().s-lef.back().f);
+	ll best = INF;
+	for (int i = 0; i < N-1; ++i)
+		best = min(best,(ll)(v[i].f-v[0].f)*(lef[i].s-lef[i].f)
+			+(ll)(v.back().f-v[i+1].f)*(rig[i+1].s-rig[i+1].f));
+	ans = max(ans,area-best);
 }
 
 int main() {
-	setIO("split"); re(N);
-	v.rsz(N); re(v);
+	setIO("split");
+	cin >> N;
+	v.resize(N);
+	for (pi& t : v) cin >> t.f >> t.s;
 	tri();
-	trav(t,v) swap(t.f,t.s);
+	for (pi& t : v) swap(t.f,t.s);
 	tri();
-	ps(ans);
+	cout << ans;
+	return 0;
 }
 ```
-
-## Note - Weak Test Data
-
-The above codes give different outputs on the following test case:
-
-```
-17
-0 0
-0 1
-0 2
-1 0
-1 1
-1 2
-2 0
-2 1
-2 2
```

Comment:
If we take 9 first points and 8 last points to 2 enclosure, we will get the optimal answer @bqi343

Reply:
I thought it was wrong (?). I will revert it soon.